File: windows/src/SecureVox.Core/Models/Recording.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace SecureVox.Core.Models;

/// <summary>
/// A voice recording with optional transcription
/// </summary>
public class Recording
{
    [Key]
    public Guid Id { get; set; } = Guid.NewGuid();

    [Required]
    [MaxLength(500)]
    public string Title { get; set; } = string.Empty;

    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
    public DateTime UpdatedAt { get; set; } = DateTime.UtcNow;

    /// <summary>
    /// Duration in seconds
    /// </summary>
    public double Duration { get; set; }

    [MaxLength(260)]
    public string? AudioFileName { get; set; }

    public long AudioFileSize { get; set; }

    public TranscriptionStatus TranscriptionStatus { get; set; } = TranscriptionStatus.Pending;

    /// <summary>
    /// Transcription progress (0.0 - 1.0)
    /// </summary>
    public double TranscriptionProgress { get; set; }

    public SourceType SourceType { get; set; } = SourceType.Recorded;

    public bool IsFavorite { get; set; }

    public bool IsDeleted { get; set; }

    public DateTime? DeletedAt { get; set; }

    // Transcription metadata
    [MaxLength(100)]
    public string? TranscriptionModel { get; set; }

    [MaxLength(100)]
    public string? TranscriptionEngine { get; set; }

    [MaxLength(10)]
    public string? DetectedLanguage { get; set; }

    [MaxLength(1000)]
    public string? TranscriptionError { get; set; }

    [MaxLength(260)]
    public string? OriginalFileName { get; set; }

    [MaxLength(10)]
    public string? Language { get; set; }

    // Navigation property - segments
    public ICollection<TranscriptSegment> Segments { get; set; } = new List<TranscriptSegment>();

    // Computed properties (not mapped to database)
    [NotMapped]
    public string FullTranscript
    {
        get
        {
            if (Segments == null || !Segments.Any())
                return string.Empty;

            return string.Join(" ", Segments
                .OrderBy(s => s.StartTime)
                .Select(s => s.Text));
        }
    }

    [NotMapped]
    public string FormattedDuration
    {
        get
        {
            var hours = (int)Duration / 3600;
            var minutes = ((int)Duration % 3600) / 60;
            var seconds = (int)Duration % 60;

            return hours > 0
                ? $"{hours}:{minutes:D2}:{seconds:D2}"
                : $"{minutes}:{seconds:D2}";
        }
    }

    [NotMapped]
    public string FormattedDate
    {
        get
        {
            var now = DateTime.Now;
            var created = CreatedAt.ToLocalTime();

            if (created.Date == now.Date)
                return $"Today {created:HH:mm}";
            if (created.Date == now.Date.AddDays(-1))
                return $"Yesterday {created:HH:mm}";
            if (created.Year == now.Year)
                return created.ToString("MMM d, HH:mm");

            return created.ToString("MMM d, yyyy HH:mm");
        }
    }

    [NotMapped]
    public string FormattedFileSize
    {
        get
        {
            string[] sizes = { "B", "KB", "MB", "GB" };
            double len = AudioFileSize;
            int order = 0;
            while (len >= 1024 && order < sizes.Length - 1)
            {
                order++;
                len /= 1024;
            }
            return $"{len:0.##} {sizes[order]}";
        }
    }

    [NotMapped]
    public bool HasAudio => !string.IsNullOrEmpty(AudioFileName) && AudioFileSize > 0;

    [NotMapped]
    public bool HasTranscript => Segments != null && Segments.Any();

    [NotMapped]
    public int? DaysUntilPermanentDeletion
    {
        get
        {
            if (!IsDeleted || DeletedAt == null)
                return null;

            // Default retention period of 30 days
            const int retentionDays = 30;
            var expirationDate = DeletedAt.Value.AddDays(retentionDays);
            var daysRemaining = (expirationDate - DateTime.UtcNow).Days;
            return Math.Max(0, daysRemaining);
        }
    }

    [NotMapped]
    public string? AudioFilePath
    {
        get
        {
            if (string.IsNullOrEmpty(AudioFileName))
                return null;

            var localFolder = Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData);
            return Path.Combine(localFolder, "SecureVox", "Recordings", AudioFileName);
        }
    }
}

File: macos/SecureVox/Models/Recording.swift
import Foundation
import SwiftData

/// A voice recording with optional transcription
@Model
final class Recording {

    // MARK: - Properties

    @Attribute(.unique) var id: UUID
    var title: String
    var createdAt: Date
    var updatedAt: Date
    var duration: TimeInterval
    var audioFileName: String?
    var audioFileSize: Int64
    var transcriptionStatus: TranscriptionStatus
    var transcriptionProgress: Double
    var sourceType: SourceType
    var isFavorite: Bool
    var isDeleted: Bool
    var deletedAt: Date?

    // Transcription metadata
    var transcriptionModel: String?
    var transcriptionEngine: String?
    var detectedLanguage: String?
    var transcriptionError: String?
    var originalFileName: String?
    var language: String?

    // MARK: - Relationships

    @Relationship(deleteRule: .cascade, inverse: \TranscriptSegment.recording)
    var segments: [TranscriptSegment]?

    // MARK: - Computed Properties

    var fullTranscript: String {
        guard let segments = segments, !segments.isEmpty else { return "" }
        return segments
            .sorted { $0.startTime < $1.startTime }
            .map { $0.text }
            .joined(separator: " ")
    }

    var formattedDuration: String {
        let hours = Int(duration) / 3600
        let minutes = (Int(duration) % 3600) / 60
        let seconds = Int(duration) % 60

        if hours > 0 {
            return String(format: "%d:%02d:%02d", hours, minutes, seconds)
        } else {
            return String(format: "%d:%02d", minutes, seconds)
        }
    }

    var formattedDate: String {
        let formatter = DateFormatter()
        let calendar = Calendar.current

        if calendar.isDateInToday(createdAt) {
            formatter.dateFormat = "'Today' HH:mm"
        } else if calendar.isDateInYesterday(createdAt) {
            formatter.dateFormat = "'Yesterday' HH:mm"
        } else if calendar.isDate(createdAt, equalTo: Date(), toGranularity: .year) {
            formatter.dateFormat = "MMM d, HH:mm"
        } else {
            formatter.dateFormat = "MMM d, yyyy HH:mm"
        }

        return formatter.string(from: createdAt)
    }

    var formattedFileSize: String {
        ByteCountFormatter.string(fromByteCount: audioFileSize, countStyle: .file)
    }

    var hasAudio: Bool {
        audioFileName != nil && audioFileSize > 0
    }

    var hasTranscript: Bool {
        guard let segments = segments else { return false }
        return !segments.isEmpty
    }

    var daysUntilPermanentDeletion: Int? {
        guard isDeleted, let deletedAt = deletedAt else { return nil }
        // Read directly from UserDefaults to avoid MainActor isolation issues
        let retentionDays = UserDefaults.standard.integer(forKey: "recycleBinRetentionDays")
        guard retentionDays > 0 else { return nil }

        let expirationDate = Calendar.current.date(byAdding: .day, value: retentionDays, to: deletedAt) ?? Date()
        let days = Calendar.current.dateComponents([.day], from: Date(), to: expirationDate).day ?? 0
        return max(0, days)
    }

    var audioURL: URL? {
        guard let audioFileName = audioFileName else { return nil }
        let documentsURL = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
        return documentsURL
            .appendingPathComponent(AppConstants.Storage.recordingsDirectory)
            .appendingPathComponent(audioFileName)
    }

    // MARK: - Initialization

    init(
        id: UUID = UUID(),
        title: String,
        createdAt: Date = Date(),
        updatedAt: Date = Date(),
        duration: TimeInterval = 0,
        audioFileName: String? = nil,
        audioFileSize: Int64 = 0,
        transcriptionStatus: TranscriptionStatus = .pending,
        transcriptionProgress: Double = 0,
        sourceType: SourceType = .recorded,
        isFavorite: Bool = false,
        isDeleted: Bool = false,
        deletedAt: Date? = nil,
        language: String? = nil,
        segments: [TranscriptSegment]? = nil
    ) {
        self.id = id
        self.title = title
        self.createdAt = createdAt
        self.updatedAt = updatedAt
        self.duration = duration
        self.audioFileName = audioFileName
        self.audioFileSize = audioFileSize
        self.transcriptionStatus = transcriptionStatus
        self.transcriptionProgress = transcriptionProgress
        self.sourceType = sourceType
        self.isFavorite = isFavorite
        self.isDeleted = isDeleted
        self.deletedAt = deletedAt
        self.language = language
        self.segments = segments
    }
}

// MARK: - Transcription Status

enum TranscriptionStatus: String, Codable {
    case pending = "pending"
    case inProgress = "in_progress"
    case completed = "completed"
    case failed = "failed"

    var displayName: String {
        switch self {
        case .pending: return "Pending"
        case .inProgress: return "Transcribing..."
        case .completed: return "Completed"
        case .failed: return "Failed"
        }
    }

    var icon: String {
        switch self {
        case .pending: return "clock"
        case .inProgress: return "waveform"
        case .completed: return "checkmark.circle"
        case .failed: return "exclamationmark.triangle"
        }
    }
}

// MARK: - Source Type

enum SourceType: String, Codable {
    case recorded = "recorded"
    case imported = "imported"
    case meeting = "meeting"
    case quick = "quick"

    var displayName: String {
        switch self {
        case .recorded: return "Recorded"
        case .imported: return "Imported"
        case .meeting: return "Meeting"
        case .quick: return "Quick"
        }
    }

    var icon: String {
        switch self {
        case .recorded: return "mic"
        case .imported: return "square.and.arrow.down"
        case .meeting: return "video"
        case .quick: return "bolt"
        }
    }
}
